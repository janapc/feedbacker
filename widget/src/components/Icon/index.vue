<template>
  <component :is="name" v-bind="$props" />
</template>

<script>
import Loading from './Loading.vue';
import Chat from './Chat.vue';
import Close from './Close.vue';
import ArrowRight from './ArrowRight.vue';
import Check from './Check.vue';
import Attention from './Attention.vue';

export default {
  components: {
    Loading,
    Chat,
    Close,
    ArrowRight,
    Check,
    Attention
  },
  props: {
    name: { type: String, required: true }
  }
};
</script>
