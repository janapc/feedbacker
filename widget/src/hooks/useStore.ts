import Store, { StoreState } from '../store';

export default function useStore (): StoreState {
  return Store;
}
