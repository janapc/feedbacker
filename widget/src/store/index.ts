import { reactive, readonly } from 'vue';

export type StoreState = {
  currentComponent: string;
  message: string;
  feedbackType: string;
  fingerprint: string;
  apiKey: string;
  currentPage: string;
};

const initialState: StoreState = {
  currentComponent: 'SelectFeedbackType',
  message: '',
  feedbackType: '',
  fingerprint: '',
  apiKey: '',
  currentPage: ''
};

const state = reactive<StoreState>({ ...initialState });

export function setCurrentComponent (component: string): void {
  state.currentComponent = component;
}

export function setMessage (message: string): void {
  state.message = message;
}

export function setFeedbackType (type: string): void {
  state.feedbackType = type;
}

export function setCurrentPage (page: string): void {
  state.currentPage = page;
}

export function setApiKey (apiKey: string): void {
  state.apiKey = apiKey;
}

export function setFingerprint (fingerprint: string): void {
  state.fingerprint = fingerprint;
}

export function resetStore (): void {
  setCurrentComponent(initialState.currentComponent);
  setMessage(initialState.message);
  setFeedbackType(initialState.feedbackType);
  setCurrentPage(initialState.currentPage);
  setApiKey(initialState.apiKey);
  setFingerprint(initialState.fingerprint);
}

export default readonly(state);
