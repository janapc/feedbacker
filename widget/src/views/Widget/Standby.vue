<template>
  <div
    id="open-box"
    class="fixed z-50 bottom-0 right-0 mb-5 mr-5 bg-brand-main rounded-full py-3 px-5 flex items-center shadow-xl cursor-pointer select-none animate__animated animate__fadeInUp animate__faster"
    @click="() => emit('open-box')"
  >
    <icon name="Chat" color="white" size="27" class="mr-3" />
    <span class="font-black text-white text-xl">Deixe um feedback</span>
  </div>
</template>

<script lang="ts">
import { defineComponent, SetupContext } from 'vue';

import Icon from '../../components/Icon/index.vue';

interface SetupReturn {
  emit: SetupContext['emit'];
}

export default defineComponent({
  components: {
    Icon
  },
  emits: ['open-box'],
  setup (_, { emit }: SetupContext): SetupReturn {
    return { emit };
  }
});
</script>
