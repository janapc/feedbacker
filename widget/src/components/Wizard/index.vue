<template>
  <component
    :is="store.currentComponent"
    @select-feedback-type="handleSelectFeedbackType"
    @next="next"
  />
</template>

<script lang="ts">
import { defineComponent } from 'vue';

import SelectFeedbackType from './SelectFeedbackType.vue';
import WriteAFeedback from './WriteAFeedback.vue';
import Success from './Success.vue';
import ErrorState from './Error.vue';

import useStore from '../../hooks/useStore';
import useNavigation, { UseNavigation } from '../../hooks/useNavigation';

import { setFeedbackType, StoreState } from '../../store';

interface SetupReturn {
  store: StoreState;
  next: UseNavigation['next'];
  handleSelectFeedbackType(type: string): void;
}

export default defineComponent({
  components: { SelectFeedbackType, WriteAFeedback, Success, Error: ErrorState },
  setup (): SetupReturn {
    const store = useStore();
    const { next } = useNavigation();

    function handleSelectFeedbackType (type: string): void {
      setFeedbackType(type);
    }

    return {
      store,
      handleSelectFeedbackType,
      next
    };
  }
});
</script>
