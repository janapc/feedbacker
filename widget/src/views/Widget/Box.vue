<template>
  <div class="box animate__animated animate_fadeInUp animate__faster" id="box">
    <header
      :class="{
        'justify-between': canShowAdditionalControlAndInfo,
        'justify-end': !canShowAdditionalControlAndInfo
      }"
      class="relative w-full flex"
    >
      <button
        id="btn-goBack"
        v-if="canShowAdditionalControlAndInfo"
        @click="back"
        :disabled="canGoBack"
        :class="{ invisible: canGoBack }"
        class="text-xl text-gray-800 focus:outline-none"
      >
        <icon name="ArrowRight" :color="colors.gray['800']" />
      </button>

      <p
        id="label"
        v-if="canShowAdditionalControlAndInfo"
        class="text-xl font-black text-center text-brand-main"
      >
        {{ label }}
      </p>

      <button
        id="close-box"
        @click="() => emit('close-box')"
        class="text-xl text-gray-800 focus:outline-none"
      >
        <icon size="14" name="Close" :color="colors.gray['800']" />
      </button>
    </header>

    <wizard />

    <footer
      class="text-gray-800 text-sm flex"
      v-if="canShowAdditionalControlAndInfo"
    >
      <icon class="mr-1" name="Chat" :color="brandColors.graydark" />
      widget by
      <span class="ml-1 font-bold">feedbacker</span>
    </footer>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, ComputedRef, SetupContext } from 'vue';
import colors from 'tailwindcss/colors';
import { brand } from '../../../palette';

import useStore from '../../hooks/useStore';
import useNavigation, { UseNavigation } from '../../hooks/useNavigation';

import Icon from '../../components/Icon/index.vue';
import Wizard from '../../components/Wizard/index.vue';

interface SetupReturn {
  emit: SetupContext['emit'];
  label: ComputedRef<string>;
  canGoBack: ComputedRef<boolean>;
  canShowAdditionalControlAndInfo: ComputedRef<boolean>;
  brandColors: Record<string, string>;
  colors: Record<string, string>;
  back: UseNavigation['back'];
}

export default defineComponent({
  emits: ['close-box'],
  components: { Icon, Wizard },
  setup (_, { emit }: SetupContext): SetupReturn {
    const store = useStore();
    const { back } = useNavigation();

    const label = computed<string>(() => {
      if (store.feedbackType === 'ISSUE') {
        return 'Reporte um problema';
      }
      if (store.feedbackType === 'IDEA') {
        return 'Nos false a sua idéia';
      }

      if (store.feedbackType === 'OTHER') {
        return 'Abra sua mente';
      }
      return 'O que você tem em mente?';
    });
    const canGoBack = computed<boolean>(() => {
      return store.currentComponent === 'SelectFeedbackType';
    });

    const canShowAdditionalControlAndInfo = computed<boolean>(() => {
      return (
        store.currentComponent !== 'Success' &&
        store.currentComponent !== 'Error'
      );
    });

    return {
      emit,
      label,
      canGoBack,
      canShowAdditionalControlAndInfo,
      colors,
      brandColors: brand,
      back
    };
  }
});
</script>

<style lang="postcss" scoped>
.box {
  @apply fixed z-50 bottom-0 right-0 mb-5 mr-5 bg-white rounded-xl py-5 px-5 flex flex-col items-center shadow-xl select-none;
  width: 400px;
}
</style>
