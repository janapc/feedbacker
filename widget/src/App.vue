<template>
  <widget />
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import Widget from './views/Widget/index.vue';

export default defineComponent({
  components: { Widget }
});
</script>
