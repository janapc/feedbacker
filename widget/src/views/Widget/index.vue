<template>
  <teleport to="body">
    <component
      @open-box="handleOpenBox"
      @close-box="handleCloseBox"
      :is="state.component"
    />
  </teleport>
</template>

<script lang="ts">
import { defineComponent, reactive, watch } from 'vue';

import useIframeControl from '../../hooks/useIframeControl';

import Standby from './Standby.vue';
import Box from './Box.vue';

import useStore from '../../hooks/useStore';

type State = {
  component: string;
};

interface SetupReturn {
  state: State;
  handleOpenBox(): void;
  handleCloseBox(): void;
}

export default defineComponent({
  components: { Standby, Box },
  setup (): SetupReturn {
    const iframeControl = useIframeControl();

    const state = reactive<State>({
      component: 'Standby'
    });

    const store = useStore();

    watch(
      () => store.currentComponent,
      () => {
        iframeControl.updateCoreValuesOnStore();
      }
    );

    function handleOpenBox (): void {
      iframeControl.notifyOpen();
      state.component = 'Box';
    }

    function handleCloseBox (): void {
      iframeControl.notifyClose();
      state.component = 'Standby';
    }

    return {
      state,
      handleOpenBox,
      handleCloseBox
    };
  }
});
</script>
