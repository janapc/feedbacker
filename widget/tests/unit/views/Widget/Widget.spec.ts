import { mount } from '@vue/test-utils';

import Widget from '../../../../src/views/Widget/index.vue';
import Standby from '../../../../src/views/Widget/Standby.vue';
import Box from '../../../../src/views/Widget/Box.vue';

import Wizard from '../../../../src/components/Wizard/index.vue';
import Icon from '../../../../src/components/Icon/index.vue';

import { setCurrentComponent, setFeedbackType } from '../../../../src/store';
import useStore from '../../../../src/hooks/useStore';

jest.mock('../../../../src/hooks/useIframeControl', () => () => ({
  updateCoreValuesOnStore: jest.fn(),
  notifyOpen: jest.fn(),
  notifyClose: jest.fn()
}));

jest.mock('../../../../src/hooks/useNavigation', () => () => ({
  back: () => {
    setFeedbackType('');
    setCurrentComponent('SelectFeedbackType');
  }
}));

describe('<Widget />', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });
  it('Should render Widget', () => {
    const wrapper = mount(Widget);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should render Widget and render the component Standby', () => {
    const wrapper = mount(Widget);
    const StandbyComponent = wrapper.findComponent(Standby);

    expect(StandbyComponent.exists()).toBeTruthy();
    expect(StandbyComponent.findComponent(Icon).props().name).toBe('Chat');
    expect(StandbyComponent.text()).toContain('Deixe um feedback');
  });

  it('Should render Widget,contain the component Standby and open the box', async () => {
    const wrapper = mount(Widget);

    expect(wrapper.vm.state.component).toBe('Standby');
    expect(wrapper.findComponent(Standby).exists()).toBeTruthy();

    await wrapper
      .findComponent(Standby)
      .find('#open-box')
      .trigger('click');

    await wrapper.vm.handleOpenBox();

    const BoxComponent = wrapper.findComponent(Box);
    expect(BoxComponent.exists()).toBeTruthy();
  });

  it('Should render Widget,contain the component Box', async () => {
    setFeedbackType('ISSUE');

    const wrapper = mount(Widget);

    await wrapper.vm.handleOpenBox();

    const BoxComponent = wrapper.findComponent(Box);

    expect(BoxComponent.exists()).toBeTruthy();
    expect(BoxComponent.find('#label').text()).toBe('Reporte um problema');
    expect(BoxComponent.findComponent(Wizard).exists()).toBeTruthy();

    expect(BoxComponent.find('#btn-goBack').attributes().class).toContain(
      'invisible'
    );
  });

  it('Should render Widget,contain the component Box and go back component', async () => {
    setFeedbackType('ISSUE');
    setCurrentComponent('writeAFeedback');

    const wrapper = mount(Widget);
    const store = useStore();

    await wrapper.vm.handleOpenBox();

    const BoxComponent = wrapper.findComponent(Box);

    expect(BoxComponent.exists()).toBeTruthy();
    expect(BoxComponent.find('#btn-goBack').attributes().class).not.toContain(
      'invisible'
    );

    await BoxComponent.find('#btn-goBack').trigger('click');

    expect(store.currentComponent).toBe('SelectFeedbackType');
    expect(store.feedbackType).toBe('');
  });

  it('Should render Widget,contain the component Box and close box', async () => {
    setFeedbackType('ISSUE');
    setCurrentComponent('writeAFeedback');
    const wrapper = mount(Widget);

    await wrapper.vm.handleOpenBox();
    expect(wrapper.vm.state.component).toBe('Box');
    const BoxComponent = wrapper.findComponent(Box);
    expect(BoxComponent.exists()).toBeTruthy();

    BoxComponent
      .find('#close-box')
      .trigger('click');

    await wrapper.vm.handleCloseBox();
    expect(wrapper.vm.state.component).toBe('Standby');
  });
});
