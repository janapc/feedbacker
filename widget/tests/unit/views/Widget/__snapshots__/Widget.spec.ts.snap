// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Widget /> Should render Widget 1`] = `
<!--teleport start-->
<!--teleport end-->
`;
